src/xenia/patcher/patcher.cc

```diff
@@ -0,0 +1,217 @@
+                                          const uint64_t hash) {
+  for (const PatchFileEntry& patchFile : loaded_patch_files) {
+    if (patchFile.title_id != title_id) {
+      continue;
+    }
+    if (patchFile.hash != 0 && patchFile.hash != hash) {
+      continue;
+    }
+
+    for (const PatchInfoEntry& patchEntry : patchFile.patch_info) {
+      if (!patchEntry.is_enabled) {
+        continue;
+      }
+      XELOGE("Applying patch for: {}({:08X}) - {}", patchFile.title_name,
+             patchFile.title_id, patchEntry.patch_name);
+      ApplyPatch(memory, &patchEntry);
+    }
+  }
+}
+
+void PatchingSystem::ApplyPatch(Memory* memory, const PatchInfoEntry* patch) {
+  for (const PatchDataEntry& patch_data_entry : patch->patch_data) {
+    uint32_t old_address_protect = 0;
+    auto address = memory->TranslateVirtual(patch_data_entry.memory_address_);
+    auto heap = memory->LookupHeap(patch_data_entry.memory_address_);
+    if (!heap) {
+      continue;
+    }
+
+    heap->QueryProtect(patch_data_entry.memory_address_, &old_address_protect);
+
+    heap->Protect(patch_data_entry.memory_address_,
+                  patch_data_entry.alloc_size_,
+                  kMemoryProtectRead | kMemoryProtectWrite);
+
+    switch (patch_data_entry.alloc_size_) {
+      case 1:
+        xe::store_and_swap(address, uint8_t(patch_data_entry.new_value_));
+        break;
+      case 2:
+        xe::store_and_swap(address, uint16_t(patch_data_entry.new_value_));
+        break;
+      case 4:
+        xe::store_and_swap(address, uint32_t(patch_data_entry.new_value_));
+        break;
+      case 8:
+        xe::store_and_swap(address, uint64_t(patch_data_entry.new_value_));
+        break;
+      default:
+        XELOGE("Unsupported patch allocation size");
+        break;
+    }
+    // Restore previous protection
+    heap->Protect(patch_data_entry.memory_address_,
+                  patch_data_entry.alloc_size_, old_address_protect);
+
+    if (!is_any_patch_applied_) {
+      is_any_patch_applied_ = true;
+    }
+  }
+}
+
+PatchFileEntry PatchingSystem::ReadPatchFile(
+    const std::filesystem::path& file_path) {
+  PatchFileEntry patchFile;
+  std::shared_ptr<cpptoml::table> patch_toml_fields;
+
+  try {
+    patch_toml_fields = cpptoml::parse_file(file_path.u8string());
+  } catch (...) {
+    XELOGE("Cannot load patch file: {}", file_path.u8string());
+    patchFile.title_id = -1;
+    return patchFile;
+  };
+
+  auto title_name = patch_toml_fields->get_as<std::string>("title_name");
+  auto title_id = patch_toml_fields->get_as<std::string>("title_id");
+  auto title_hash = patch_toml_fields->get_as<std::string>("hash");
+
+  patchFile.title_id = strtoul((*title_id).c_str(), NULL, 16);
+  patchFile.hash = strtoull((*title_hash).c_str(), NULL, 16);
+  patchFile.title_name = *title_name;
+
+  auto tarr = patch_toml_fields->get_table_array("patch");
+
+  for (auto table : *tarr) {
```

Comment:
If going with other proposed variable name changes. `table` should probably be `patch_table`?